subprojects/core/src/main/java/org/gradle/execution/plan/DefaultExecutionPlan.java

```diff
@@ -333,13 +333,47 @@ public NodeInVisitingSegment apply(TaskNode taskNode) {
     }
 
     @Override
-    public Set<Task> getDependencies(Task task) {
-        TaskNode node = nodeMapping.get(task);
-        ImmutableSet.Builder<Task> builder = ImmutableSet.builder();
-        for (Node dependencyNode : node.getDependencySuccessors()) {
-            dependencyNode.collectTaskInto(builder);
-        }
-        return builder.build();
+    public ExecutionDependencies getDependencies(Task task) {
+        return getDependencies(nodeMapping.get(task));
+    }
+
+    @Override
+    public ExecutionDependencies getDependencies(TransformationNodeIdentifier transformation) {
+        return getDependencies(nodeMapping.get(transformation));
+    }
+
+    private ExecutionDependencies getDependencies(final Node node) {
+        return new ExecutionDependencies() {
+            @Override
+            public Set<Task> getTasks() {
+                final ImmutableSet.Builder<Task> builder = ImmutableSet.builder();
+                collectDependencies(node, new Node.Visitor() {
+                    @Override
+                    public void visitTask(Task task) {
+                        builder.add(task);
+                    }
+                });
+                return builder.build();
+            }
+
+            @Override
+            public Set<TransformationNodeIdentifier> getTransformations() {
+                final ImmutableSet.Builder<TransformationNodeIdentifier> builder = ImmutableSet.builder();
+                collectDependencies(node, new Node.Visitor() {
+                    @Override
+                    public void visitTransformation(TransformationNodeIdentifier transformation) {
+                        builder.add(transformation);
+                    }
+                });
+                return builder.build();
+            }
```

Comment:
When we turn this into a visitor, it would be nice to have some symmetry. Currently for tasks we get a full `Task` instance, while for transformations we only get the identity. I think it should be okay to get the identity in both cases.

Reply:
That would seriously complicate things in `DefaultTaskExecutionGraph`, though.

Reply:
Using a visitor or visiting `WorkIdentifier`s?

Reply:
Visiting `WorkIdentifier`s, because then the corresponding tasks would have to be looked up again.

Reply:
OK, so let's visit the `Node`s. It's easy to get to `Task` instances from a `LocalTaskNode`.